File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/IsSymmetric.java
package com.only4.algorithm.version4j.leetcode;

import com.only4.algorithm.version4j.extra.TreeNode;

public class IsSymmetric {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) return true;
        return isMirror(root.left, root.right);
    }

    private boolean isMirror(TreeNode leftNode, TreeNode rightNode) {
        if (leftNode == null && rightNode == null) return true;
        if (leftNode == null || rightNode == null) return false;

        return leftNode.val == rightNode.val
                && isMirror(leftNode.left, rightNode.right)
                && isMirror(leftNode.right, rightNode.left);
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/isSymmetric.kt
package com.only4.algorithm.version4k.leetcode

import com.only4.algorithm.version4k.extra.TreeNode

fun isSymmetric(root: TreeNode?): Boolean {
    if (root == null) return true
    return isMirror(root.left, root.right)
}

private fun isMirror(leftNode: TreeNode?, rightNode: TreeNode?): Boolean {
    if (leftNode == null && rightNode == null) return true
    if (leftNode == null || rightNode == null) return false

    return leftNode.`val` == rightNode.`val`
            && isMirror(leftNode.left, rightNode.right)
            && isMirror(leftNode.right, rightNode.left)
}
